Core: Show "booting" progress until story is specified or errors

tmeasday opened the pull request:
Issue: https://github.com/storybookjs/storybook/issues/20250

## What I did

Show the progress bar on the preview component until we learn the story is specified or an error occurs.

If preview.js takes a long time to load, the story specified event will not fire until it is done.

<img width="1624" alt="Screenshot 2022-12-28 at 9 02 47 pm" src="https://user-images.githubusercontent.com/132554/209796418-cfd1714e-8a50-4e1a-8ae0-4ca2a8202631.png">

A (potential) downside here is you *always* see the above briefly while the SB is loading.

## How to test

Add a "speed loop" to a preview.js, then load the SB:

```
const start = new Date();
while (new Date() - start < 2000);
```

tmeasday committed:
Show a "booting" progress message until the story is specified

(Or an error occurs).

JReinhold commented on code/ui/manager/src/components/preview/preview.tsx:
```diff
@@ -90,15 +92,21 @@ const createCanvas = (id: string, baseUrl = 'iframe.html', withLoader = true): A
             : (!storiesFailed && !storiesConfigured) || rootLoading;
+          const isBooting = !storySpecified && !storiesFailed;
 
+          console.log({ withLoader, isLoading, isBooting, progress });
```
```suggestion
```

JReinhold commented on code/ui/manager/src/components/preview/preview.tsx:
```diff
@@ -90,15 +92,21 @@ const createCanvas = (id: string, baseUrl = 'iframe.html', withLoader = true): A
+                        <Loader
+                          id="preview-loader"
+                          role="progressbar"
+                          progress={progress || { value: 1, message: 'Booting' }}
```
```suggestion
                          progress={progress || { value: 1, message: 'Loading' }}
```

Wouldn't "Loading" be more user friendly? Or "Fetching modules" would be more precise I guess in this instance, but that might be too much.

I just feel like Booting is something that an OS or server does, not really what our manager is doing here.

JReinhold reviewed (changes requested):
I think this is a good first step, makes it way better at least.

But in general I'm not a fan of showing a progress bar to users when progress can't be determined, but that's a minor thing. We probably need an indetermined loading spinner or similar in general in our UI.

yannbf commented:
> A (potential) downside here is you always see the above briefly while the SB is loading.

What if you added a timeout to only show this after like 100-200ms?

tmeasday commented:
> We probably need an indetermined loading spinner or similar in general in our UI.

One issue we are trying to avoid is having too many different spinners[1] flash up in the early stages, so I think a better UI would be some version of a progressbar that is clearly 

[1] First we show the loading progress bar as the build is compiled, then we show this thing while `preview.js` is executed, then we show a docs skeleton or spinner while the docs/story is imported.

tmeasday committed:
Apply suggestions from code review

Co-authored-by: Jeppe Reinhold <jeppe@chromatic.com>

tmeasday committed:
Don't show a progressbar if booting

tmeasday committed:
Simplify things further

tmeasday commented:
Actually @JReinhold @yannbf I changed my mind on this. Looking closer, we already show a circular spinner for a moment when a SB first starts (before the entry loads) so I might as well keep it simple and simply augment the existing logic to wait a bit longer (for the story to be specified by the preview -- which happens after initialization -- rather than when configured).

tmeasday committed:
Simplify a little

tmeasday committed:
Simplify

yannbf commented:
Sounds good! Could you check the behavior with a composed storybook?

yannbf commented on code/ui/manager/src/components/preview/preview.tsx:
```diff
@@ -85,11 +84,12 @@ const createCanvas = (id: string, baseUrl = 'iframe.html', withLoader = true): A
             }
           }, []);
 
+          const isBooting = !storySpecified && !storiesFailed;
```
Could you add a little comment here? People might not understand what booting means

tmeasday committed:
Add some comments to explain things

tmeasday committed:
Remove `storiesConfigured` et al, and simply to just `ready`.

This is all we had for refs too, and the logic for keeping it simple is the same.

tmeasday commented:
Thanks for the prompt to check properly @yannbf. When I did so I ran into some problems and decided I was just adding some extra complexity to some code that was more complex than it needed to be already. So I simplified things to just have a `ready` state (same as we do in refs) that gets set when the story is selected or an error is emitted. That works in all the cases we are concerned about.

@ndelangen you should probably check this as the original author.

tmeasday commented on code/lib/manager-api/src/modules/stories.ts:
```diff
@@ -54,8 +55,7 @@ export interface SubState {
   storyId: StoryId;
   viewMode: ViewMode;
-  storiesConfigured: boolean;
-  storiesFailed?: Error;
```
Note we never actually looked at this value anywhere, so to simplify things I just dropped it.

yannbf replied:
Renaming the `storiesConfigured` property is a breaking change, are you sure you want to do it?

There are many projects that use it, and here are a couple: [first](https://github.com/AndrianBalanescu/builder/blob/master/packages/storybook/src/Tab.tsx#L64), [second](https://github.com/stackbit/storybook-addon/blob/13fd96d87a0ae2d2c4015cf9d446ef632f2e706c/src/Panel.jsx#L22), [third](https://github.com/jh3y/storybook-addon-studybook/blob/main/src/components/badge.js#L61)

And apparently [some](https://github.com/Andarist/react-spectrum-v3/blob/master/.storybook-v3/theme.js#L18) [addons](https://github.com/adamyonk/storybook-dark-mode/blob/master/src/Tool.tsx#L106) listen to a `storiesConfigured` event 🤔

tmeasday replied:
I guess we could put it back, but not use it ourselves. What about the `storiesErrrored` property? We didn't seem to use it but it could be that addons/etc do?

Would we deprecate them?

It seems better to just have a `ready` property and for it to be a bit vaguer as to what it means ("configured" being incorrect as it turns out).

yannbf replied:
I did not see any `storiesErrrored` property anywhere, not sure what you mean?
I'm fine with the change to `ready`, I just wanted to let you know that this will break people's code, and I wanted to know if it was intended

tmeasday replied:
Discussed and the conclusion is people should not be reaching inside our state. If they are they'll need to update in a breaking release (although we might not consider such a change breaking).

yannbf replied:
Apparently there is a user facing hook which gives access to the internal state:
https://storybook.js.org/docs/react/addons/addons-api#storybook-hooks

tmeasday committed:
Show docs immediately on first load

yannbf commented:
I was testing it out and saw that between loading states, there are "blank" states, please let me know if that is intended:

Here's a "Fast 3G" setting, cache disabled:
![2023-01-03 10 19 17](https://user-images.githubusercontent.com/1671563/210329501-4ddc4fc7-4cbe-4b47-a8d6-465c03b48251.gif)

Here's no throttling, with cache disabled and refreshing a couple times to see the behavior better (it's similar with cache as well):
![2023-01-03 10 19 52](https://user-images.githubusercontent.com/1671563/210329636-06149650-1127-416b-ac21-da8d087e0a1f.gif)

ndelangen commented:
@tmeasday I'll do a proper review of this next week.

tmeasday committed:
Don't set ready until we are actually ready!

tmeasday commented:
Ok, I improved it, although we still need to figure out the old events. Thanks for the tip @yannbf. I think this also highlights that we remove the `DOCS_PREPARING` UI before we actually have finished rendering the docs underneath. Not really noticeable in normal network conditions, but I wonder if we should fix it. In this PR or another?

ndelangen commented:
@tmeasday and I discussed this, this morning, and concluded there's a bit more refactoring needed to make this work proper.

tmeasday committed:
Change refs + stories to use a common interface

Add fields:
 - `stories` (index once loaded into a hash)
 - `indexError` (if there's an error loading that)
 - `previewInitialized` (if the preview is ready to render)

tmeasday commented:
@ndelangen @yannbf OK I overhauled this a bit, mainly based on discussions we had. What I did:

Changed the old internal state fields `storiesHash`, `storiesConfigured` and `storiesFailed` to three new fields: `stories`, `stories` (which starts `undefined`),`indexError` and `previewInitialized`.

This is a pretty big breaking change for anyone reaching inside our state (which they shouldn't do I guess). I changed the name of `storiesHash` to be consistent between the main `stories` module and `refs`. I guess I could have called it `storiesHash` and changed `refs` here instead of stories.

One side effect of this is that now we actually show errors if the main index fails. Which we might want to consider again re: https://github.com/storybookjs/storybook/issues/20044 and https://github.com/storybookjs/storybook/issues/19094 @ndelangen 

I added stories for the above ☝️.

I didn't add stories for the preview states as the preview stories need to be overhauled.

tmeasday committed:
Drop some logs

ndelangen commented:
@tmeasday Do you want me to review now? Seems there's some CI failures. Or is that on purpose?

yannbf commented:
Let me know once CI is green so I can test it out <3

tmeasday commented:
I was waiting to hear if y'all were ok with the above before fixing it. The sandboxes work if you want to give it a spin though.

tmeasday committed:
Update tests

tmeasday committed:
Missed this

tmeasday committed:
Update snapshots

tmeasday committed:
Update various tests/stories

tmeasday committed:
One more

tmeasday committed:
Merge remote-tracking branch 'origin/next' into tom/sb-1123-sb20250-vite-does-not-show-a-spinner-2

tmeasday commented:
@ndelangen any idea what's happening with [these errors](https://app.circleci.com/pipelines/github/storybookjs/storybook/37943/workflows/da86f083-9aa3-4040-903a-90d2dc9c110f/jobs/467016)?

```
../../lib/manager-api/src/lib/addons.ts(122,15): error TS2551: Property '___STORYBOOK_ADDONS_MANAGER' does not exist on type 'typeof globalThis'. Did you mean '__STORYBOOKADDONS__'?
../../lib/manager-api/src/lib/addons.ts(123,12): error TS2551: Property '___STORYBOOK_ADDONS_MANAGER' does not exist on type 'typeof globalThis'. Did you mean '__STORYBOOKADDONS__'?
../../lib/manager-api/src/lib/addons.ts(125,17): error TS2551: Property '___STORYBOOK_ADDONS_MANAGER' does not exist on type 'typeof globalThis'. Did you mean '__STORYBOOKADDONS__'?
../../lib/manager-api/src/modules/refs.ts(280,50): error TS7017: Element implicitly has an 'any' type because type 'typeof globalThis' has no index signature.
../../lib/manager-api/src/modules/release-notes.ts(7,9): error TS2339: Property 'RELEASE_NOTES_DATA' does not exist on type 'typeof globalThis'.
../../lib/manager-api/src/modules/versions.ts(10,9): error TS2339: Property 'VERSIONCHECK' does not exist on type 'typeof globalThis'.
```

yannbf commented:
@tmeasday maybe `stories` should be called `entries` instead? or something along those lines?

ndelangen committed:
Merge branch 'next' into tom/sb-1123-sb20250-vite-does-not-show-a-spinner-2

ndelangen commented:
@tmeasday I'm taking a look at those errors!

ndelangen committed:
make changes to typings files to ensure this the check command succeeds

tmeasday commented:
Thanks @ndelangen !

tmeasday committed:
Update stories->index and add back-compatibility

tmeasday committed:
Merge branch 'tom/sb-1123-sb20250-vite-does-not-show-a-spinner-2' of https://github.com/storybookjs/storybook into tom/sb-1123-sb20250-vite-does-not-show-a-spinner-2

tmeasday committed:
Various updates

tmeasday committed:
More fixes

tmeasday committed:
Fix up app stories to be initialized

tmeasday committed:
Bunch of updates I missed

tmeasday committed:
Missed one more

tmeasday committed:
Even more fixes

tmeasday committed:
Another issue

tmeasday committed:
Catch missing stories

tmeasday committed:
Add debugging to built-in sb

tmeasday commented:
@ndelangen @yannbf OK, this is ready to go I think, apart from the globals issue.

As discussed with @yannbf I:

- Changed the name from `stories/storiesHash` to `index` as it's no just stories any more.
- Added deprecated getters for the old properties, as we have a documented `useStorybookState` API (for better or worse)

ndelangen committed:
Merge pull request #20598 from storybookjs/norbert/make-typescript-check-work-valid

make changes to typings files to ensure this the check command succeeds

tmeasday commented:
Thanks for making it green! Any change of a ✅ from @ndelangen or @yannbf?

ndelangen commented:
@tmeasday will put a calender event om my monday morning to review 👍

ndelangen committed:
Merge branch 'next' into tom/sb-1123-sb20250-vite-does-not-show-a-spinner-2

ndelangen committed:
when testing for regressions, this didn't show up at all. renaming it fixed it
